# 07 — Disparity metric (standardized mean difference)

Our primary measure of demographic sensitivity is the **standardized mean difference (SMD)** between two demographic groups within the same `(model, condition, prompt)` cell. We adopt the sign conventions

$$\text{SMD}_\text{race} = \frac{\overline{x}_\text{Black} - \overline{x}_\text{White}}{\sigma_\text{pooled}}, \quad \text{SMD}_\text{gender} = \frac{\overline{x}_\text{Female} - \overline{x}_\text{Male}}{\sigma_\text{pooled}}, \quad \text{SMD}_\text{accent} = \frac{\overline{x}_\text{non-SAE} - \overline{x}_\text{SAE}}{\sigma_\text{pooled}}$$

so a positive race SMD at a given prompt means the model produces **larger numeric responses to Black-speaker audio than to White-speaker audio** at that prompt; a positive gender SMD means **larger responses to Female than to Male**; a positive accent SMD means **larger responses to non-SAE than to SAE speakers**. SAE = participants whose only self-reported accent is "Standard American English" (Rule A, strict — see `accent_simplified` in `participants.csv`); non-SAE = everyone else. The accent axis is expected to co-vary with race on this sample (the 81 audit-eligible speakers who self-report SAE *and* another accent are 77% Black) — we treat the two axes as complementary descriptive views, not independent signals.

Within each `(model, condition, prompt)` cell and each run $i \in \{1, 2, 3\}$ we compute the SMD; we then aggregate to the scenario level by **(i) averaging across prompts within each iteration, and (ii) averaging the per-iteration scenario means across runs** — *per-iter first, then iter-average*. Because the SMD standardizes responses within each prompt before aggregation, scenario-level averages are meaningful even when individual prompts within a scenario differ substantially in raw magnitude or unit: the standardization absorbs those differences, so averaging across prompts within a scenario combines effect sizes on a common scale.

**This notebook is the pipeline.** It runs every step end-to-end from `data/model_outputs/extracted_responses.parquet` and writes the four CSV families under `data/model_outputs/disparity/` as products of execution — no precomputed shortcuts:

1. **§ 3** — Build the SMD analysis universe (strict-audit-eligible × scenario quant × non-probe × valid response categories).
2. **§ 4** — Per-(model, prompt, axis, group) three-category validity tables (the SMD's complementary dimension) → `validity_rates_3cat__{condition}.csv`.
3. **§ 5** — Speaker-level median imputation + 50 % imputability prompt filter → `imputation_drop_report.csv`.
4. **§ 6** — The SMD primitive walked through on one real `(model, condition, prompt, run)` cell.
5. **§ 7** — Per-prompt iter-averaged SMDs at both winsorize variants → `prompt_level_smds.csv`.
6. **§ 8** — Participant-level cluster bootstrap (B = 2 000) → `scenario_smds_with_ci.csv`. **This is the long step (~75 min sequential).**

All preprocessing — imputation + prompt filter + winsorize bounds — is **estimated once on the original sample and held fixed across bootstrap iterations**, isolating SMD-estimate uncertainty from filter-membership uncertainty.

Eligibility filter: the SMD universe is gated on `strict_audit_eligible` (n = 500), matching the rest of the paper (notebooks 03 / 04 / 05 / 10 and `src/eval/speech_metrics.py`). Axes are configured in `disparity.AXES` — adding a new binary contrast (e.g., expert-labeled accent) is a one-line registration.

**Inputs:** `../data/model_outputs/extracted_responses.parquet`, `../data/metadata/{participants,prompts}.csv`  
**Outputs (written by this notebook):** `../data/model_outputs/disparity/{imputation_drop_report.csv, validity_rates_3cat__{condition}.csv, prompt_level_smds.csv, scenario_smds_with_ci.csv}`

## Setup

In [1]:
import sys
import time
from pathlib import Path

import numpy as np
import pandas as pd

REPO_ROOT = Path.cwd().parent
sys.path.insert(0, str(REPO_ROOT / "src"))
from eval import disparity

META = REPO_ROOT / "data" / "metadata"
MO   = REPO_ROOT / "data" / "model_outputs"
DISP = MO / "disparity"
DISP.mkdir(parents=True, exist_ok=True)

## 1. Preprocessing rules

Two preprocessing steps are needed before SMDs are well-defined; both are estimated **once on the original sample** and reused across all bootstrap iterations.

**Speaker-level median imputation.** When a `(model, condition, prompt, participant)` cell has at least one valid numeric run, missing runs (refusals, INVALIDs, upstream API errors) are filled with the participant's own median across their valid iterations. Cells with no valid run at all are dropped. This makes the SMD primitive defined in cells where a model occasionally fails *stochastically* across runs rather than *systematically*. § 5 will show that this materially affects `gpt-audio-1.5`, whose failure profile is qualitatively different from Gemini's and Qwen's.

**50 % imputability prompt filter.** After imputation, we retain a prompt only if at least 50 % of strict-audit-eligible participants who saw it have at least one valid numeric response. Prompts where the model's behavior is too degenerate to support a meaningful comparison are dropped. The threshold is `disparity.IMPUTABILITY_THRESHOLD = 0.50`.

**Winsorization.** Per-prompt quantile clipping at `[pct, 1 − pct]`. The published headline uses `pct = 0.01` (1 % winsorize); we also produce the raw (no-winsorize) variant. Quantile bounds are estimated once on the original sample.

**Cluster bootstrap (B = 2 000, seed = 42).** 95 % percentile CIs from a participant-level cluster bootstrap: each iteration resamples strict-audit-eligible participants with replacement and re-runs the SMD computation using the resampled multiplicities as observation weights. Both the imputation rule and the 50 % prompt filter stay fixed across iterations.

All constants live in `src/eval/disparity.py` as the single source of truth:

In [2]:
print("Pipeline constants (src/eval/disparity.py):")
print(f"  IMPUTABILITY_THRESHOLD = {disparity.IMPUTABILITY_THRESHOLD}")
print(f"  WINSORIZE_MODES        = {disparity.WINSORIZE_MODES}")
print(f"  N_BOOT                 = {disparity.N_BOOT}")
print(f"  BOOT_SEED              = {disparity.BOOT_SEED}")
print(f"  CI                     = {disparity.CI}")
print(f"  PROBE_QIDS  (excluded) = {disparity.PROBE_QIDS}")
print(f"  BP_QID      (excluded) = {disparity.BP_QID}") # blood-pressure prompt excluded due to nature of expected output

Pipeline constants (src/eval/disparity.py):
  IMPUTABILITY_THRESHOLD = 0.5
  WINSORIZE_MODES        = (None, 0.01)
  N_BOOT                 = 2000
  BOOT_SEED              = 42
  CI                     = 0.95
  PROBE_QIDS  (excluded) = (72, 73)
  BP_QID      (excluded) = 25


## 2. Build the SMD analysis universe

Steps:

1. Load `extracted_responses.parquet`, drop within-`task_id` duplicates (first wins).
2. Derive `participant_id` from `clip_id` (e.g., `P0001_q01` → `P0001`); canonical-text rows have `clip_id = None` and get filtered out downstream.
3. Merge demographics from `participants.csv` (race, gender, strict-audit eligibility).
4. Merge prompt metadata from `prompts.csv` (scenario, prompt type, is_quantitative).
5. Coerce `extracted_value` → numeric `value_num` (NaN for non-`number` rows).
6. Three-category collapse: `disparity.classify_3cat(extracted_kind, error_type)` → `{numeric, invalid, refusal, missing_data}`.
7. `disparity.build_smd_universe()` applies the universe filter: strict-audit-eligible × scenario × quantitative × non-probe × `q25` excluded × `response_3cat ∈ {numeric, invalid, refusal}`.

In [3]:
extr = pd.read_parquet(MO / "extracted_responses.parquet")
n_raw = len(extr)
extr = extr.drop_duplicates(subset="task_id", keep="first").reset_index(drop=True)
print(f"extracted_responses: {n_raw:,} raw rows → {len(extr):,} after dedup")

participants = pd.read_csv(META / "participants.csv")
prompts      = pd.read_csv(META / "prompts.csv")

extr["participant_id"] = extr["clip_id"].astype(str).str.split("_").str[0]
extr.loc[extr["clip_id"].isna(), "participant_id"] = None

df = extr.merge(
    participants[["participant_id",
                   "race_simplified", "gender_simplified", "accent_simplified",
                   "strict_audit_eligible"]],
    on="participant_id", how="left",
)

evaluable_qids = set(prompts[
    ((prompts["prompt_type"] == "scenario") & prompts["is_quantitative"])
    | prompts["question_id"].isin([72, 73])
]["question_id"].astype(float))
df["question_id"]    = pd.to_numeric(df["question_id"], errors="coerce")
df["is_quantitative"] = df["question_id"].isin(evaluable_qids)
df = df.merge(
    prompts[["question_id", "scenario", "prompt_type"]].rename(
        columns={"prompt_type": "pt_canonical"}),
    on="question_id", how="left",
)

def _to_num(v, k):
    if k != "number" or v is None or (isinstance(v, float) and pd.isna(v)):
        return np.nan
    try:
        return float(v)
    except (TypeError, ValueError):
        return np.nan
df["value_num"]     = [_to_num(v, k) for v, k in zip(df["extracted_value"], df["extracted_kind"])]
df["response_3cat"] = [disparity.classify_3cat(k, e) for k, e in zip(df["extracted_kind"], df["error_type"])]

universe = disparity.build_smd_universe(df, prompts)
print(f"\nSMD analysis universe: {len(universe):,} rows  |  unique participants: {universe['participant_id'].nunique()}")
print("per (provider, condition):")
print(universe.groupby(["provider", "condition"]).size().to_string())
print("\nresponse_3cat distribution in the universe:")
print(universe["response_3cat"].value_counts().to_string())
print("\naxes configured for the SMD computation (sign = pos − neg):")
for a in disparity.AXES:
    print(f"  {a.name:7s}  pos={a.pos!r:11s}  neg={a.neg!r:11s}  column={a.col}")

extracted_responses: 261,900 raw rows → 261,900 after dedup


src/eval/disparity.py:117: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  & df["strict_audit_eligible"].fillna(False).astype(bool)



SMD analysis universe: 233,424 rows  |  unique participants: 500
per (provider, condition):
provider  condition                   
gemini    direct_audio_response           37851
          external_transcript_response    39957
          self_transcript_response        39957
openai    direct_audio_response           37851
qwen      direct_audio_response           37851
          external_transcript_response    39957

response_3cat distribution in the universe:
response_3cat
numeric    175822
invalid     40818
refusal     16784

axes configured for the SMD computation (sign = pos − neg):
  race     pos='Black'      neg='White'      column=race_simplified
  gender   pos='Female'     neg='Male'       column=gender_simplified
  accent   pos='non-SAE'    neg='SAE'        column=accent_simplified


## 3. Validity tables — the SMD's complementary dimension

The SMD is defined **only over valid numeric responses**: it compares the distribution of model outputs for group $a$ against group $b$ *conditional on both producing a numeric answer*. If one speaker group systematically elicits more invalid, refused, or missing responses than another — a form of differential model behavior in its own right — that disparity is **invisible to the SMD**.

We surface this complementary dimension as a per-(model, prompt, axis, group) three-category rate table. The per-prompt CSV is written per condition; below we also pool across prompts to summarize and compute the (Black − White) and (Female − Male) gaps within each rate column.

In [4]:
validity_tables = {}
for cond in sorted(universe["condition"].unique()):
    vt = disparity.build_validity_table(universe, cond)
    if vt.empty:
        continue
    validity_tables[cond] = vt
    out = DISP / f"validity_rates_3cat__{cond}.csv"
    vt.to_csv(out, index=False)
    print(f"wrote {out.relative_to(REPO_ROOT)}  ({len(vt):,} rows)")

vr = validity_tables["direct_audio_response"]
pooled = (vr.groupby(["model_id", "axis", "group"])
             .agg(n_prompts=("question_id", "nunique"),
                  mean_pct_numeric=("pct_numeric", "mean"),
                  mean_pct_invalid=("pct_invalid", "mean"),
                  mean_pct_refusal=("pct_refusal", "mean")).round(3).reset_index())
print("\nMean three-category rates per (model × axis × group), direct_audio, pooled across prompts:")
pooled

wrote data/model_outputs/disparity/validity_rates_3cat__direct_audio_response.csv  (972 rows)


wrote data/model_outputs/disparity/validity_rates_3cat__external_transcript_response.csv  (648 rows)
wrote data/model_outputs/disparity/validity_rates_3cat__self_transcript_response.csv  (324 rows)



Mean three-category rates per (model × axis × group), direct_audio, pooled across prompts:


,model_id,axis,group,n_prompts,mean_pct_numeric,mean_pct_invalid,mean_pct_refusal
0,Qwen/Qwen2.5-Omni-7B,accent,SAE,54,0.879,0.121,0.000
1,Qwen/Qwen2.5-Omni-7B,accent,non-SAE,54,0.882,0.117,0.000
2,Qwen/Qwen2.5-Omni-7B,gender,Female,54,0.877,0.123,0.000
3,Qwen/Qwen2.5-Omni-7B,gender,Male,54,0.885,0.114,0.000
4,Qwen/Qwen2.5-Omni-7B,race,Black,54,0.880,0.120,0.001
5,Qwen/Qwen2.5-Omni-7B,race,White,54,0.883,0.117,0.000
6,gemini-3.1-flash-lite-preview,accent,SAE,54,0.648,0.349,0.003
7,gemini-3.1-flash-lite-preview,accent,non-SAE,54,0.658,0.337,0.004
8,gemini-3.1-flash-lite-preview,gender,Female,54,0.649,0.346,0.005
9,gemini-3.1-flash-lite-preview,gender,Male,54,0.660,0.337,0.003


In [5]:
def axis_gap(vr_df, axis, group_pos, group_neg, rate_col):
    pos = vr_df[(vr_df["axis"] == axis) & (vr_df["group"] == group_pos)][
        ["model_id", "question_id", rate_col]].rename(columns={rate_col: "pos"})
    neg = vr_df[(vr_df["axis"] == axis) & (vr_df["group"] == group_neg)][
        ["model_id", "question_id", rate_col]].rename(columns={rate_col: "neg"})
    j = pos.merge(neg, on=["model_id", "question_id"])
    j["diff"] = j["pos"] - j["neg"]
    return j.groupby("model_id")["diff"].mean().round(4)

gap_cols = {}
for ax in disparity.AXES:
    sign = f"{ax.pos}−{ax.neg}"
    for rate in ("numeric", "refusal", "invalid"):
        gap_cols[f"{rate}  {sign}"] = axis_gap(vr, ax.name, ax.pos, ax.neg, f"pct_{rate}")
gap = pd.DataFrame(gap_cols)
print("Mean within-axis gap in three-category rates per model, direct_audio.")
print("Sign convention matches the SMD: positive = pos-group − neg-group within each axis.")
gap

Mean within-axis gap in three-category rates per model, direct_audio.
Sign convention matches the SMD: positive = pos-group − neg-group within each axis.


,numeric Black−White,refusal Black−White,invalid Black−White,numeric Female−Male,refusal Female−Male,invalid Female−Male,numeric non-SAE−SAE,refusal non-SAE−SAE,invalid non-SAE−SAE
model_id,,,,,,,,,
Qwen/Qwen2.5-Omni-7B,-0.0037,0.0006,0.0031,-0.0083,0.0003,0.0080,0.0034,0.0005,-0.0039
gemini-3.1-flash-lite-preview,0.0225,0.0023,-0.0247,-0.0108,0.0015,0.0093,0.0098,0.0014,-0.0112
gpt-audio-1.5,0.0053,-0.0037,-0.0016,-0.0022,0.0063,-0.0041,0.0023,-0.0025,0.0002


## 4. Imputation + 50 % prompt filter (drop report)

Two diagnostics per `(model, condition)` cell:

- `pct_dropped` — fraction of prompts the model handled too degenerately for the SMD to be meaningful (no valid numeric response from ≥ 50 % of speakers).
- `fraction_imputed_among_kept` — within the retained prompts, the fraction of `(participant, prompt, run)` cells whose value came from speaker-level median imputation rather than from a numeric response on that specific run.

These two numbers tell **opposite stories** about a model's failure profile.

In [6]:
drop_rows = []
imputed = {}
for cond in sorted(universe["condition"].unique()):
    for m in sorted(universe.loc[universe["condition"] == cond, "model_id"].unique()):
        imp_kept, drop = disparity.impute_and_filter(universe, m, cond)
        imputed[(m, cond)] = imp_kept
        drop_rows.append({"model_id": m, "condition": cond, **drop})
drop_df = pd.DataFrame(drop_rows)
out = DISP / "imputation_drop_report.csv"
drop_df[["model_id", "condition", "n_prompts_total", "n_prompts_kept",
           "n_prompts_dropped", "pct_dropped", "fraction_imputed_among_kept"]].to_csv(out, index=False)
print(f"wrote {out.relative_to(REPO_ROOT)}\n")
drop_df[["model_id", "condition", "n_prompts_total", "n_prompts_kept",
           "n_prompts_dropped", "pct_dropped", "fraction_imputed_among_kept",
           "dropped_qids"]]

wrote data/model_outputs/disparity/imputation_drop_report.csv



,model_id,condition,n_prompts_total,n_prompts_kept,n_prompts_dropped,pct_dropped,fraction_imputed_among_kept,dropped_qids
0,Qwen/Qwen2.5-Omni-7B,direct_audio_response,54,49,5,9.26,0.000000,"[5, 15, 27, 32, 66]"
1,gemini-3.1-flash-lite-preview,direct_audio_response,54,37,17,31.48,0.005243,"[5, 10, 13, 15, 19, 20, 23, 27, 32, 34, 35, 40..."
2,gpt-audio-1.5,direct_audio_response,54,52,2,3.70,0.282215,"[19, 50]"
3,Qwen/Qwen2.5-Omni-7B,external_transcript_response,54,36,18,33.33,0.000000,"[1, 3, 4, 5, 7, 15, 17, 26, 27, 32, 35, 39, 50..."
4,gemini-3.1-flash-lite-preview,external_transcript_response,54,45,9,16.67,0.001744,"[5, 27, 32, 34, 50, 51, 58, 59, 67]"
5,gemini-3.1-flash-lite-preview,self_transcript_response,54,53,1,1.85,0.000329,[27]


**Reading the drop report — `gpt-audio-1.5`'s qualitatively different failure profile.**

On `direct_audio_response`:

- **Gemini** has `pct_dropped ≈ 31.5 %` (17 / 54 prompts excluded) but `fraction_imputed_among_kept ≈ 0.5 %`. Read together: Gemini tends to fail **systematically** on whole prompts — emitting `INVALID` consistently across runs and participants — so those prompts get dropped wholesale, but the prompts that survive the filter come through with virtually no per-run gaps to impute.
- **Qwen** has `pct_dropped ≈ 9.3 %` and `fraction_imputed_among_kept ≈ 0.0 %`. Same shape as Gemini at a smaller scale: prompts either work or don't, with essentially no stochastic per-run failures.
- **`gpt-audio-1.5`** is different: `pct_dropped ≈ 3.7 %` (only 2 prompts excluded) but `fraction_imputed_among_kept ≈ 28.2 %`. Roughly **one in three** of `gpt-audio-1.5`'s kept `(participant, prompt, run)` cells was a stochastic per-run failure that the imputation rule rescued by filling in the participant's median across their other valid runs. Without that rule, the SMD primitive would frequently see groups with effective sample size < 2 in a `(prompt, run)` cell for `gpt-audio-1.5` and return NaN, biasing the headline scenario aggregations toward Gemini and Qwen.

Translated to the audit framing: **Gemini and Qwen fail by *refusing*; `gpt-audio-1.5` fails by flaking.** The two failure modes are differently invisible to the SMD — refusal-style failures get exposed by the `pct_dropped` column; flake-style failures get absorbed by imputation. Tracking both numbers per model is essential to read an SMD across the three models on equal footing.

## 5. The SMD primitive — walked through on a real cell

To make `disparity.smd_weighted()` concrete, we walk through it on one real `(model, condition, prompt, run)` cell from the imputed dataframe. We pick `gpt-audio-1.5 × direct_audio × q01 × run_index=1` (chosen because gpt-audio's high imputation rate stresses the primitive), pull the post-imputation `value_imputed` arrays for Black-speaker and White-speaker rows, and call `smd_weighted` directly. We then call `disparity.per_iter_prompt_smds()` over the whole `(model, condition)` cell and confirm the row for `(q01, run_index=1)` carries the same number — i.e. the high-level function reduces to the primitive on this real data point.

In [7]:
MODEL_DEMO, COND_DEMO = "gpt-audio-1.5", "direct_audio_response"
QID_DEMO, RUN_DEMO = 1.0, 1

imp_demo = imputed[(MODEL_DEMO, COND_DEMO)]
cell = imp_demo[(imp_demo["question_id"] == QID_DEMO) & (imp_demo["run_index"] == RUN_DEMO)]

black = cell.loc[cell["race_simplified"] == "Black",  "value_imputed"].to_numpy()
white = cell.loc[cell["race_simplified"] == "White",  "value_imputed"].to_numpy()
wB = np.ones_like(black, dtype=int)
wW = np.ones_like(white, dtype=int)

print(f"Cell: {MODEL_DEMO} × {COND_DEMO} × q{int(QID_DEMO):02d} × run_index={RUN_DEMO}")
print(f"  n_Black = {len(black)}  mean = {black.mean():.3f}  sd = {black.std(ddof=1):.3f}")
print(f"  n_White = {len(white)}  mean = {white.mean():.3f}  sd = {white.std(ddof=1):.3f}")

smd = disparity.smd_weighted(black, white, wB, wW)
print(f"\n  disparity.smd_weighted(Black, White, 1s, 1s) = {smd:+.6f}")

Cell: gpt-audio-1.5 × direct_audio_response × q01 × run_index=1
  n_Black = 113  mean = 4.770  sd = 3.329
  n_White = 103  mean = 4.568  sd = 1.722

  disparity.smd_weighted(Black, White, 1s, 1s) = +0.075183


In [8]:
# Cross-check: call the per-(prompt, iter) function on the whole (model, condition) cell
# and pull the row for (q01, run_index=1).
weights_unit = {p: 1 for p in imp_demo["participant_id"].unique()}
pi = disparity.per_iter_prompt_smds(imp_demo, value_col="value_imputed", weights=weights_unit)
row = pi[(pi["question_id"] == QID_DEMO) & (pi["run_index"] == RUN_DEMO)].iloc[0]
print(f"per_iter_prompt_smds row: smd_race = {row['smd_race']:+.6f}")
assert abs(row["smd_race"] - smd) < 1e-12, "primitive vs per-iter mismatch"
print("✓ disparity.smd_weighted(...) and disparity.per_iter_prompt_smds(...) agree byte-for-byte on this real cell.")

per_iter_prompt_smds row: smd_race = +0.075183


✓ disparity.smd_weighted(...) and disparity.per_iter_prompt_smds(...) agree byte-for-byte on this real cell.


## 6. Prompt-level SMDs

For each `(model, condition, winsorize_pct)` cell, we run the atomic per-`(prompt, iter)` SMD computation and then iter-average across the three runs to get one per-prompt SMD per axis. The result is written to `prompt_level_smds.csv` — the intermediate layer for inspecting whether a scenario-level SMD is being pulled by one or two extreme prompts. Both winsorize variants (raw and 1 %) are produced.

In [9]:
prompt_smds_rows = []
for (m, cond), imp in imputed.items():
    if len(imp) == 0:
        continue
    audit_pids = sorted(imp["participant_id"].unique())
    weights_unit = {p: 1 for p in audit_pids}
    for w in disparity.WINSORIZE_MODES:
        imp_w = disparity.winsorize_per_prompt(imp, w)
        pi = disparity.per_iter_prompt_smds(imp_w, value_col="value_w", weights=weights_unit)
        ps = disparity.prompt_iter_avg_smds(pi)
        ps.insert(0, "winsorize_pct", w)
        ps.insert(0, "condition",     cond)
        ps.insert(0, "model_id",      m)
        prompt_smds_rows.append(ps)

prompt_smds = pd.concat(prompt_smds_rows, ignore_index=True)
out = DISP / "prompt_level_smds.csv"
prompt_smds.to_csv(out, index=False)
print(f"wrote {out.relative_to(REPO_ROOT)}  ({len(prompt_smds):,} rows)")

sub = prompt_smds[prompt_smds["winsorize_pct"] == 0.01]
agg = {"n_prompts": ("question_id", "nunique")}
for ax in disparity.AXES:
    col = f"smd_{ax.name}"
    agg[f"{ax.name}_mean"] = (col, "mean")
    agg[f"{ax.name}_p10"]  = (col, lambda s: s.quantile(0.10))
    agg[f"{ax.name}_p90"]  = (col, lambda s: s.quantile(0.90))
summary = sub.groupby(["model_id", "condition"]).agg(**agg).round(3)
print("\nPer-prompt SMD summary at 1% winsorize (mean, 10th, 90th percentile), per axis:")
summary

wrote data/model_outputs/disparity/prompt_level_smds.csv  (544 rows)

Per-prompt SMD summary at 1% winsorize (mean, 10th, 90th percentile), per axis:


/tmp/ipykernel_1842512/782782602.py:16: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  prompt_smds = pd.concat(prompt_smds_rows, ignore_index=True)


n_prompts  \
model_id                      condition                                 
Qwen/Qwen2.5-Omni-7B          direct_audio_response                49   
                              external_transcript_response         36   
gemini-3.1-flash-lite-preview direct_audio_response                37   
                              external_transcript_response         45   
                              self_transcript_response             53   
gpt-audio-1.5                 direct_audio_response                52   

                                                            race_mean  \
model_id                      condition                                 
Qwen/Qwen2.5-Omni-7B          direct_audio_response             0.042   
                              external_transcript_response      0.064   
gemini-3.1-flash-lite-preview direct_audio_response            -0.009   
                              external_transcript_response     -0.003   
                              self_transcript_response         -0.011   
gpt-audio-1.5                 direct_audio_response             0.031   

                                                            race_p10  \
model_id                      condition                                
Qwen/Qwen2.5-Omni-7B          direct_audio_response           -0.237   
                              external_transcript_response    -0.252   
gemini-3.1-flash-lite-preview direct_audio_response           -0.348   
                              external_transcript_response    -0.330   
                              self_transcript_response        -0.239   
gpt-audio-1.5                 direct_audio_response           -0.148   

                                                            race_p90  \
model_id                      condition                                
Qwen/Qwen2.5-Omni-7B          direct_audio_response            0.307   
                              external_transcript_response     0.269   
gemini-3.1-flash-lite-preview direct_audio_response            0.292   
                              external_transcript_response     0.275   
                              self_transcript_response         0.248   
gpt-audio-1.5                 direct_audio_response            0.201   

                                                            gender_mean  \
model_id                      condition                                   
Qwen/Qwen2.5-Omni-7B          direct_audio_response              -0.017   
                              external_transcript_response       -0.058   
gemini-3.1-flash-lite-preview direct_audio_response              -0.044   
                              external_transcript_response       -0.006   
                              self_transcript_response           -0.046   
gpt-audio-1.5                 direct_audio_response              -0.010   

                                                            gender_p10  \
model_id                      condition                                  
Qwen/Qwen2.5-Omni-7B          direct_audio_response             -0.202   
                              external_transcript_response      -0.268   
gemini-3.1-flash-lite-preview direct_audio_response             -0.281   
                              external_transcript_response      -0.192   
                              self_transcript_response          -0.210   
gpt-audio-1.5                 direct_audio_response             -0.163   

                                                            gender_p90  \
model_id                      condition                                  
Qwen/Qwen2.5-Omni-7B          direct_audio_response              0.189   
                              external_transcript_response       0.208   
gemini-3.1-flash-lite-preview direct_audio_response              0.146   
                              external_transcript_response       0.182   
                              self_transcript_response           0.102   
gpt-audio-1.5           

## 7. Cluster bootstrap → scenario-level SMDs with 95 % CIs

The headline output. For each `(model, condition, winsorize_pct)` cell we run `disparity.summarize_with_ci()` with `n_boot = 2000` and `seed = 42`. Each bootstrap iteration:

1. Samples strict-audit-eligible participants with replacement → frequency weights $w_p$.
2. Per `(prompt, iter)`: frequency-weighted SMD for every axis in `disparity.AXES`.
3. Per iter: scenario-mean across prompts (NaN-skip).
4. Per scenario: mean across the (≤ 3) per-iter scenario means.

95 % percentile CIs come from the resulting bootstrap distribution per `(scenario, axis)`.

**Parallelism.** The 12 `(model × condition × winsorize)` cells are independent, so we run them with a `ProcessPoolExecutor` over up to 12 workers. Each worker re-runs `summarize_with_ci()` on its own cell; the seed is the same per cell, so results are byte-deterministic regardless of worker scheduling. On a 96-core machine this drops the wallclock from ~2 h sequential to ~10 min (set by the slowest cell). Result written to `data/model_outputs/disparity/scenario_smds_with_ci.csv`.

In [10]:
from concurrent.futures import ProcessPoolExecutor, as_completed


def _bootstrap_cell(model_id, condition, winsorize_pct, imp):
    """Worker: run summarize_with_ci for one (model_id, condition, winsorize_pct) cell.

    Defined at notebook-top so ProcessPoolExecutor can pickle it. Re-imports
    `disparity` inside the worker process so module-level constants (AXES, N_BOOT,
    etc.) are picked up from src/eval/disparity.py.
    """
    import sys
    from pathlib import Path
    sys.path.insert(0, str(Path.cwd().parent / "src"))
    from eval import disparity as _disp

    imp_w = _disp.winsorize_per_prompt(imp, winsorize_pct)
    sc = _disp.summarize_with_ci(
        imp_w, value_col="value_w",
        n_boot=_disp.N_BOOT, seed=_disp.BOOT_SEED, ci=_disp.CI,
    )
    sc.insert(0, "winsorize_pct", winsorize_pct)
    sc.insert(0, "condition",     condition)
    sc.insert(0, "model_id",      model_id)
    return sc


jobs = [(m, cond, w, imp)
         for (m, cond), imp in imputed.items() if len(imp) > 0
         for w in disparity.WINSORIZE_MODES]
print(f"{len(jobs)} (model × condition × winsorize) cells to bootstrap "
      f"with B={disparity.N_BOOT}; using up to {min(len(jobs), 12)} worker processes.")

scenario_rows = []
t0 = time.time()
with ProcessPoolExecutor(max_workers=min(len(jobs), 12)) as ex:
    futures = {ex.submit(_bootstrap_cell, m, cond, w, imp): (m, cond, w)
                for (m, cond, w, imp) in jobs}
    for n_done, fut in enumerate(as_completed(futures), 1):
        m, cond, w = futures[fut]
        sc = fut.result()
        scenario_rows.append(sc)
        wlabel = "raw" if w is None else f"wins{int(w * 100)}"
        elapsed = time.time() - t0
        print(f"  [{n_done:2d}/{len(jobs)}  {elapsed/60:5.1f} min wall] {m} / {cond} ({wlabel})")

scenario_smds = pd.concat(scenario_rows, ignore_index=True).sort_values(
    ["model_id", "condition", "winsorize_pct", "scenario", "axis"]
).reset_index(drop=True)
out = DISP / "scenario_smds_with_ci.csv"
scenario_smds.to_csv(out, index=False)
print(f"\nwrote {out.relative_to(REPO_ROOT)}  ({len(scenario_smds):,} rows)  in {(time.time()-t0)/60:.1f} min wall")

12 (model × condition × winsorize) cells to bootstrap with B=2000; using up to 12 worker processes.


  [ 1/12    6.7 min wall] Qwen/Qwen2.5-Omni-7B / external_transcript_response (raw)


  [ 2/12    6.7 min wall] Qwen/Qwen2.5-Omni-7B / external_transcript_response (wins1)


  [ 3/12    6.8 min wall] gemini-3.1-flash-lite-preview / direct_audio_response (raw)


  [ 4/12    6.8 min wall] gemini-3.1-flash-lite-preview / direct_audio_response (wins1)


  [ 5/12    8.2 min wall] gemini-3.1-flash-lite-preview / external_transcript_response (raw)


  [ 6/12    8.2 min wall] gemini-3.1-flash-lite-preview / external_transcript_response (wins1)


  [ 7/12    9.0 min wall] Qwen/Qwen2.5-Omni-7B / direct_audio_response (raw)


  [ 8/12    9.0 min wall] Qwen/Qwen2.5-Omni-7B / direct_audio_response (wins1)


  [ 9/12    9.4 min wall] gpt-audio-1.5 / direct_audio_response (wins1)


  [10/12    9.5 min wall] gpt-audio-1.5 / direct_audio_response (raw)


  [11/12    9.6 min wall] gemini-3.1-flash-lite-preview / self_transcript_response (raw)


  [12/12    9.6 min wall] gemini-3.1-flash-lite-preview / self_transcript_response (wins1)

wrote data/model_outputs/disparity/scenario_smds_with_ci.csv  (396 rows)  in 9.6 min wall


/tmp/ipykernel_1842512/3444867731.py:46: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  scenario_smds = pd.concat(scenario_rows, ignore_index=True).sort_values(


## 7b. Appendix: direction-aware aggregation (`direction_clear = 1`)

The headline SMDs in § 7 are **descriptive** — a non-zero SMD means responses differ across groups, but the metric itself is silent on whether that difference is good or bad for the minority group. For many prompts the directionality is unambiguous (e.g., offering smaller dollar values in a hiring scenario is straightforwardly worse), but for many others it is not (spending more or fewer hours with one's kids is hard to label).

For an appendix view that *does* code directionality, we use the three flags newly added to `prompts.csv`:

- `direction_clear ∈ {0, 1}` — researcher consensus on whether the prompt's directionality is unambiguous enough to interpret an SMD sign. We subset to `direction_clear = 1`.
- `direction ∈ {negative, positive, neutral}` — `negative` means a more-negative SMD (under our `pos − neg` sign convention: Black − White, Female − Male, non-SAE − SAE) is the harmful direction for the minority. `positive` means the opposite.
- `scale_prompt ∈ {0, 1}` — a small set of prompts uses a 1-10 Likert scale; for the scale-and-`direction=positive` ones (e.g., "on a scale of 1 to 10, how concerned…") the natural domain-space flip is `11 − x`.

**Sign-unification.** To produce a unified "negative SMD = bad for the minority" convention across all `direction_clear = 1` prompts, we negate per-prompt SMDs for the `direction=positive` ones. For SMD specifically this is exact — the affine flip `x → c − x` (any `c`) leaves the pooled SD unchanged and flips the mean-difference, so `SMD → −SMD`. The 1 % winsorize is symmetric on both tails, so it commutes with this flip. Equivalently, we negate `value_imputed` for positive-direction prompts before the bootstrap and let the existing pipeline produce flipped SMDs naturally — no changes to `disparity.py` needed.

**Caveat.** For scale prompts the flip has a clean domain reading (`11 − x` on 1-10). For unbounded positive-direction prompts (e.g., minutes-to-fix-a-roof), the negation is a sign-convention move on the metric, not a transform of the raw response domain — we are *not* recommending negated raw values to anyone reading the responses, only relabeling the dimensionless SMD axis. The conservative alternative would be to drop the unbounded positive-direction prompts; we report the all-flipped variant here so the appendix headline includes the full `direction_clear = 1` set.

**Output.** This section writes two parallel CSVs alongside the headline outputs:

- `prompt_level_smds__direction_clear.csv` — per-prompt SMDs over the appendix subset, with sign-flip applied.
- `scenario_smds_with_ci__direction_clear.csv` — scenario-level SMDs + 95 % bootstrap CIs over the appendix subset.

Bootstrap details are identical to § 7 (B = 2 000, seed = 42, same participant-level cluster sampling). The bootstrap runs on the *same* imputed `(model, condition)` cells from § 4 — we don't re-run imputation or the imputability filter against the smaller appendix subset, because those rules are about a model's behavior on each prompt independently and shouldn't change just because we're now looking at a subset of prompts.


In [ ]:
# Build the appendix-variant imputed cells: subset to direction_clear == 1, and negate
# value_imputed for direction == positive (equivalent to SMD-negation post-winsorize).
prompts_dir = prompts.copy()
prompts_dir["question_id"] = pd.to_numeric(prompts_dir["question_id"], errors="coerce")
clear_qids    = set(prompts_dir.loc[prompts_dir["direction_clear"] == 1, "question_id"].dropna())
positive_qids = set(prompts_dir.loc[(prompts_dir["direction_clear"] == 1)
                                     & (prompts_dir["direction"] == "positive"), "question_id"].dropna())

print(f"direction_clear == 1: {len(clear_qids)} prompts (qids: {sorted(int(q) for q in clear_qids)})")
print(f"  ↳ of which direction == positive (SMDs will be sign-negated): "
      f"{len(positive_qids)} (qids: {sorted(int(q) for q in positive_qids)})")
print(f"  ↳ note: BP_QID={disparity.BP_QID} is independently excluded upstream and is not in the bootstrap input either way.")

imputed_appendix = {}
for (m, cond), imp in imputed.items():
    if len(imp) == 0:
        imputed_appendix[(m, cond)] = imp
        continue
    sub = imp[imp["question_id"].isin(clear_qids)].copy()
    flip_mask = sub["question_id"].isin(positive_qids)
    sub.loc[flip_mask, "value_imputed"] = -sub.loc[flip_mask, "value_imputed"]
    imputed_appendix[(m, cond)] = sub

# Coverage diagnostic per (model, condition): how many appendix prompts survive imputability + filter?
diag = []
for (m, cond), imp in imputed_appendix.items():
    n_q = imp["question_id"].nunique()
    n_pos = imp[imp["question_id"].isin(positive_qids)]["question_id"].nunique()
    diag.append({"model_id": m, "condition": cond,
                  "n_appendix_prompts": int(n_q),
                  "n_flipped_positive": int(n_pos),
                  "n_rows": int(len(imp))})
print("\nAppendix subset coverage per (model × condition):")
pd.DataFrame(diag)


In [ ]:
# Per-prompt SMDs over the appendix subset (both winsorize variants).
prompt_smds_rows_appendix = []
for (m, cond), imp in imputed_appendix.items():
    if len(imp) == 0:
        continue
    audit_pids = sorted(imp["participant_id"].unique())
    weights_unit = {p: 1 for p in audit_pids}
    for w in disparity.WINSORIZE_MODES:
        imp_w = disparity.winsorize_per_prompt(imp, w)
        pi = disparity.per_iter_prompt_smds(imp_w, value_col="value_w", weights=weights_unit)
        ps = disparity.prompt_iter_avg_smds(pi)
        ps.insert(0, "winsorize_pct", w)
        ps.insert(0, "condition",     cond)
        ps.insert(0, "model_id",      m)
        prompt_smds_rows_appendix.append(ps)

prompt_smds_appendix = pd.concat(prompt_smds_rows_appendix, ignore_index=True)
out = DISP / "prompt_level_smds__direction_clear.csv"
prompt_smds_appendix.to_csv(out, index=False)
print(f"wrote {out.relative_to(REPO_ROOT)}  ({len(prompt_smds_appendix):,} rows)")

sub = prompt_smds_appendix[prompt_smds_appendix["winsorize_pct"] == 0.01]
agg = {"n_prompts": ("question_id", "nunique")}
for ax in disparity.AXES:
    col = f"smd_{ax.name}"
    agg[f"{ax.name}_mean"] = (col, "mean")
    agg[f"{ax.name}_p10"]  = (col, lambda s: s.quantile(0.10))
    agg[f"{ax.name}_p90"]  = (col, lambda s: s.quantile(0.90))
summary_app = sub.groupby(["model_id", "condition"]).agg(**agg).round(3)
print("\nAppendix per-prompt SMD summary at 1% winsorize "
      "(negative = bad for minority; flipped sign applied to positive-direction prompts):")
summary_app


In [ ]:
# Cluster bootstrap → scenario-level SMDs with 95% CIs on the appendix subset.
# Same primitives, same B, same seed; only the input cells (filtered + sign-flipped) differ.

jobs_app = [(m, cond, w, imp)
             for (m, cond), imp in imputed_appendix.items() if len(imp) > 0
             for w in disparity.WINSORIZE_MODES]
print(f"{len(jobs_app)} (model × condition × winsorize) appendix cells to bootstrap "
      f"with B={disparity.N_BOOT}; using up to {min(len(jobs_app), 12)} worker processes.")

scenario_rows_app = []
t0 = time.time()
with ProcessPoolExecutor(max_workers=min(len(jobs_app), 12)) as ex:
    futures = {ex.submit(_bootstrap_cell, m, cond, w, imp): (m, cond, w)
                for (m, cond, w, imp) in jobs_app}
    for n_done, fut in enumerate(as_completed(futures), 1):
        m, cond, w = futures[fut]
        sc_ = fut.result()
        scenario_rows_app.append(sc_)
        wlabel = "raw" if w is None else f"wins{int(w * 100)}"
        elapsed = time.time() - t0
        print(f"  [{n_done:2d}/{len(jobs_app)}  {elapsed/60:5.1f} min wall] {m} / {cond} ({wlabel})")

scenario_smds_appendix = pd.concat(scenario_rows_app, ignore_index=True).sort_values(
    ["model_id", "condition", "winsorize_pct", "scenario", "axis"]
).reset_index(drop=True)
out = DISP / "scenario_smds_with_ci__direction_clear.csv"
scenario_smds_appendix.to_csv(out, index=False)
print(f"\nwrote {out.relative_to(REPO_ROOT)}  ({len(scenario_smds_appendix):,} rows)  "
      f"in {(time.time()-t0)/60:.1f} min wall")

# Quick sanity: headline view, direct_audio × wins1.
sc_app = scenario_smds_appendix
head = sc_app[(sc_app["winsorize_pct"] == 0.01) & (sc_app["condition"] == "direct_audio_response")].copy()
head["CI95"] = head.apply(
    lambda r: f"[{r['ci_lo']:+.2f}, {r['ci_hi']:+.2f}]" if pd.notna(r["ci_lo"]) else "", axis=1)
head["excludes_0"] = (head["ci_lo"] > 0) | (head["ci_hi"] < 0)
print("\n=== Appendix (direction_clear=1, signs unified) — direct_audio × wins1 ===")
print("sign convention: negative SMD ⇒ harmful direction for the minority group.")
for ax in disparity.AXES:
    print(f"\n--- axis = {ax.name} ({ax.pos} − {ax.neg}, after positive-direction sign-flip) ---")
    t = (head[head["axis"] == ax.name]
         .pivot(index="scenario", columns="model_id", values=["point", "CI95", "excludes_0"]))
    print(t.to_string())


## 8. Headline tables

Load the freshly-written `scenario_smds_with_ci.csv` and display the headline configuration (`winsorize_pct = 0.01`, `direct_audio_response`) pivoted by `(scenario × model)` for each axis. A scenario whose CI excludes zero is a scenario where the protocol detects sensitivity at the 95 % level under this aggregation — **descriptive, not a harm score** (see § 9).

In [11]:
sc = pd.read_csv(DISP / "scenario_smds_with_ci.csv")
head_da = sc[(sc["winsorize_pct"] == 0.01) & (sc["condition"] == "direct_audio_response")].copy()
head_da["CI95"]      = head_da.apply(
    lambda r: f"[{r['ci_lo']:+.2f}, {r['ci_hi']:+.2f}]" if pd.notna(r["ci_lo"]) else "", axis=1)
head_da["excludes_0"] = (head_da["ci_lo"] > 0) | (head_da["ci_hi"] < 0)

for ax in disparity.AXES:
    print(f"\n=== Scenario-level SMDs — direct_audio_response (winsorize=1%), axis={ax.name} ===")
    print(f"sign convention: {ax.name} = {ax.pos} − {ax.neg}")
    t = (head_da[head_da["axis"] == ax.name]
         .pivot(index="scenario", columns="model_id", values=["point", "CI95", "excludes_0"]))
    print(t.to_string())


=== Scenario-level SMDs — direct_audio_response (winsorize=1%), axis=race ===
sign convention: race = Black − White
                                              point                                                             CI95                                                         excludes_0                                            
model_id                       Qwen/Qwen2.5-Omni-7B gemini-3.1-flash-lite-preview gpt-audio-1.5 Qwen/Qwen2.5-Omni-7B gemini-3.1-flash-lite-preview   gpt-audio-1.5 Qwen/Qwen2.5-Omni-7B gemini-3.1-flash-lite-preview gpt-audio-1.5
scenario                                                                                                                                                                                                                           
Caring Household Members                     0.1198                     -0.096844      0.030195       [-0.01, +0.25]                [-0.22, +0.03]  [-0.07, +0.12]                False                

In [12]:
# Same headline view for the transcript-based conditions where the model ran them.
for cond in ["external_transcript_response", "self_transcript_response"]:
    sub = sc[(sc["winsorize_pct"] == 0.01) & (sc["condition"] == cond)].copy()
    if sub.empty:
        continue
    sub["CI95"] = sub.apply(
        lambda r: f"[{r['ci_lo']:+.2f}, {r['ci_hi']:+.2f}]" if pd.notna(r["ci_lo"]) else "", axis=1)
    for ax in disparity.AXES:
        print(f"\n=== {cond} (winsorize=1%, axis={ax.name}: {ax.pos} − {ax.neg}) ===")
        print(sub[sub["axis"] == ax.name]
              .pivot(index="scenario", columns="model_id", values=["point", "CI95"]).to_string())


=== external_transcript_response (winsorize=1%, axis=race: Black − White) ===
                                              point                                               CI95                              
model_id                       Qwen/Qwen2.5-Omni-7B gemini-3.1-flash-lite-preview Qwen/Qwen2.5-Omni-7B gemini-3.1-flash-lite-preview
scenario                                                                                                                            
Caring Household Members                   0.274296                      0.007883       [+0.05, +0.48]                [-0.12, +0.12]
Civic and Religious Activities             0.013243                     -0.112497       [-0.11, +0.14]                [-0.24, +0.01]
Educational Activities                          NaN                     -0.059616                                     [-0.24, +0.11]
Finance                                    0.127063                      0.126652       [-0.01, +0.24]                [+0.0

In [13]:
# Count scenarios whose 95% CI excludes zero per (model × condition × axis).
excl = sc[sc["winsorize_pct"] == 0.01].copy()
excl["excludes_0"] = (excl["ci_lo"] > 0) | (excl["ci_hi"] < 0)
ct = (excl.groupby(["model_id", "condition", "axis"])
         .agg(n_scenarios=("scenario", "nunique"),
              n_excludes_0=("excludes_0", "sum")).reset_index())
ct["frac_excludes_0"] = (ct["n_excludes_0"] / ct["n_scenarios"]).round(3)
print("Scenarios whose 95 % CI excludes zero, per (model × condition × axis):")
ct

Scenarios whose 95 % CI excludes zero, per (model × condition × axis):


,model_id,condition,axis,n_scenarios,n_excludes_0,frac_excludes_0
0,Qwen/Qwen2.5-Omni-7B,direct_audio_response,accent,11,1,0.091
1,Qwen/Qwen2.5-Omni-7B,direct_audio_response,gender,11,1,0.091
2,Qwen/Qwen2.5-Omni-7B,direct_audio_response,race,11,3,0.273
3,Qwen/Qwen2.5-Omni-7B,external_transcript_response,accent,11,2,0.182
4,Qwen/Qwen2.5-Omni-7B,external_transcript_response,gender,11,2,0.182
5,Qwen/Qwen2.5-Omni-7B,external_transcript_response,race,11,3,0.273
6,gemini-3.1-flash-lite-preview,direct_audio_response,accent,11,2,0.182
7,gemini-3.1-flash-lite-preview,direct_audio_response,gender,11,1,0.091
8,gemini-3.1-flash-lite-preview,direct_audio_response,race,11,4,0.364
9,gemini-3.1-flash-lite-preview,external_transcript_response,accent,11,1,0.091


## 9. Interpretation caveats

The SMD measures whether responses to comparable spoken requests differ across speaker groups; **it does not, on its own, say which direction is harmful**. For some prompts the directionality is unambiguous — a model offering smaller dollar values to Black or female speakers in a hiring scenario is straightforwardly worse — but for many it is not. If a model assigns higher numeric estimates to Black-speaker audio on a household-time prompt, whether that constitutes a harmful stereotype, a benign association, or an artifact of the prompt's framing is a judgment the metric cannot make. We accordingly treat scenario-aggregated SMDs as **descriptive** rather than as a harm score: they signal *where the protocol detects sensitivity*, not whether that sensitivity is itself harmful.

A further limitation, reiterated: the SMD is defined **only over valid numeric responses**: it compares the distribution of model outputs for group $a$ against group $b$ conditional on both producing a numeric answer. The validity-rate tables in § 3 are the complementary dimension that surfaces compliance asymmetries the SMD cannot see, and should be read alongside any SMD interpretation.

## 10. Headline summary

- **SMD = (mean_a − mean_b) / pooled_sd** within each `(model, condition, prompt, run)` cell, with sign conventions race = Black − White, gender = Female − Male, accent = non-SAE − SAE.
- **Axes are configured in `disparity.AXES`**: race, gender, accent. Each axis is a single binary contrast; adding another (e.g., expert-labeled accent) is a one-line registration.
- **Scenario aggregation: per-iter first, then iter-average** — preserves NaN-skip semantics correctly across the three runs.
- **Preprocessing held fixed across bootstrap iterations**: speaker-level median imputation, 50 % imputability prompt filter, 1 % winsorize.
- **Failure-profile asymmetry across models** (§ 4): Gemini and Qwen drop whole prompts when they fail; `gpt-audio-1.5` flakes per-run and gets rescued by speaker-level imputation (~28 % of `gpt-audio-1.5` cells are imputed values vs ~0 – 0.5 % for the other two).
- **Complementary metric**: per-(model, prompt, axis, group) three-category validity rates surface compliance asymmetries the SMD cannot see.
- **Uncertainty**: 95 % percentile CIs from a participant-level cluster bootstrap with B = 2 000, seed = 42, held fixed across `(model, condition, winsorize)` cells.
- **Accent-race collinearity**: the accent axis is *not* independent of race on this sample (recruitment was stratified on race × gender, not accent), so the accent SMD should be read as a complementary descriptive view of the race SMD rather than a separate signal.
- **Tables, not harm scores**: scenario-level SMDs describe where the protocol detects sensitivity; directionality interpretation is left to the reader.

**Artifacts written by this notebook** under `data/model_outputs/disparity/`:
- `validity_rates_3cat__{direct_audio,external_transcript,self_transcript}_response.csv` — three-category validity tables (the SMD complement), per axis × group
- `imputation_drop_report.csv` — per-cell drop rates + imputation fractions
- `prompt_level_smds.csv` — intermediate per-prompt SMDs (both winsorize variants), one column per axis
- `scenario_smds_with_ci.csv` — the headline scenario-level SMDs with 95 % CIs, one row per (scenario × axis)